### PR TITLE
feat: create Gov.UK Account provider

### PR DESCRIPTION
**Pull request by jerotire**
| Q                | A
| ---------------- | ---
| Type             | Feature
| Ticket           | https://dvsa.atlassian.net/browse/VOL-3384

**Description of the change:**
Introduces GOV.UK Account provider for PHP League OAuth2-Client.
- Supports self configuration from OIDC .well-known endpoint.
- Supports CachedKeySet for caching public JWKs.

https://oauth2-client.thephpleague.com/

https://www.sign-in.service.gov.uk/documentation

### CODE REVIEW

**Comment by jerotire** [comment]
Security analysis fails when uploading to GitHub. Repo needs to be public.

**Comment by sdh100shaun** [comment]
> Security analysis fails when uploading to GitHub. Repo needs to be public.

add a conditional to the action, not critical it is uploaded

File: src/Provider/GovUkAccount.php [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+        $grant,
+        array $options = []
+    ): \Dvsa\GovUkAccount\Token\AccessToken {
+        $issuedAt = new DateTimeImmutable();
+        $expiryDelta = $options['access_token_expiry_delta'] ??
+            static::DEFAULT_ACCESS_TOKEN_EXPIRY;
+        $expiresAt = $issuedAt->modify($expiryDelta);
+
+        if (!$expiresAt instanceof DateTimeImmutable) {
+            throw new \InvalidArgumentException("Could not create expiresAt using a delta on issuedAt.");
```
Comment (fibble) [inline src/Provider/GovUkAccount.php]:
```suggestion
            throw new \InvalidArgumentException('Could not create expiresAt using a delta on issuedAt.');
```

File: src/Provider/GovUkAccount.php [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+    private function parseIdentityAssuranceKey($jwk): Key
+    {
+        if (!is_array($jwk)) {
+            $jwk = json_decode($jwk);
+        }
+
+        $key = JWK::parseKey($jwk);
+        if (empty($key)) {
+            throw new InvalidArgumentException(
+                "Unable to create KEY object from identity_assurance_public_key"
```
Comment (fibble) [inline src/Provider/GovUkAccount.php]:
```suggestion
                'Unable to create KEY object from identity_assurance_public_key'
```

File: src/Exception/ApiException.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+namespace Dvsa\GovUkAccount\Exception;
+
+use Throwable;
+
+class ApiException extends GovUkAccountException
+{
+    public array $data;
+
+    public function __construct(
+        string    $message = "",
```
Comment (fibble) [inline src/Exception/ApiException.php]:
```suggestion
        string    $message = '',
```

**Review by fibble — CHANGES_REQUESTED** [review]
v.minor string quote suggestions.

**Comment by jerotire** [comment]
> > Security analysis fails when uploading to GitHub. Repo needs to be public.
> 
> add a conditional to the action, not critical it is uploaded

@sdh100shaun how important is it if the uploading of the sarif file fails? There is already a conditional in the action to ensure it is always run -- however I have just added "continue-on-error". Meaning the step won't go red (and therefore the workflow) if it cannot upload the sarif file. I was hoping for a middle ground (some Yellow status). What you think so far? 

      - name: Upload result to GitHub Code Scanning
        if: ${{ success() || failure() }}
        continue-on-error: true
        uses: github/codeql-action/upload-sarif@v2
        with:
          sarif_file: snyk.sarif

**Comment by sdh100shaun** [comment]
> > > Security analysis fails when uploading to GitHub. Repo needs to be public.
> > 
> > 
> > add a conditional to the action, not critical it is uploaded
> 
> @sdh100shaun how important is it if the uploading of the sarif file fails? There is already a conditional in the action to ensure it is always run -- however I have just added "continue-on-error". Meaning the step won't go red (and therefore the workflow) if it cannot upload the sarif file. I was hoping for a middle ground (some Yellow status). What you think so far?
> 
> ```
>   - name: Upload result to GitHub Code Scanning
>     if: ${{ success() || failure() }}
>     continue-on-error: true
>     uses: github/codeql-action/upload-sarif@v2
>     with:
>       sarif_file: snyk.sarif
> ```

Not critical continue on error seems ok

**Review by fibble — APPROVED** [review]

**Review by sdh100shaun — APPROVED** [review]

**Pull request merged**